text/0000-async-await.md

```diff
@@ -0,0 +1,171 @@
+- Feature Name: Support async / await syntax sugar on top of promises
+- Start Date: 2019-03-02
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Promises are very useful to facilitate the management of asynchronous behaviour, but working with them can quickly become tedious and error-prone.
+
+Many programming language communities (JavaScript, Dart, Rust, C#, F#,Python, Scala, Kotlin) have recognized this problem and converged on the async/await construct.
+
+# Motivation
+
+Pony is quite different from most programming languages that are widely used by today's programmers. For this reason, Pony already offers many difficult challenges which may put off a large amount of programmers who could otherwise become active members of the community and contribute to the language's success.
+
+One of the difficulties in Pony is no doubt working with asynchronous behaviours. Promises help a lot in that regard, but their usage can easily degenerate into what some people have termed the "callback hell", which makes it much harder to reason about the flow of a program.
+
+The async / await pattern makes working with asynchronous code, and specially reasoning about it, much easier. Given that many programmers are already familiar with the concept, introducing it to Pony should help ease the learning curve of the language quite significantly without impacting negatively its overall design and performance.
+
+# Detailed design
+
+The async / await pattern can be implemented as syntax sugar by the compiler because the Pony Programming Language already supports:
+
+* lambdas which can capture some of the scope surrounding it (explicitly).
+* asynchronous computation that uses lambdas to continue the flow of execution (as promises do in Pony) at some point in the future.
+
+The reason why these pre-conditions are enought is that they allow the automated translation of code as follows, (assuming `async fun` are functions that can use behaviours to return a value computed asynchronously, in the form of a `Promise`):
+
+```
+async fun lambdaV(t: T): Promise[V] => ...
+async fun lambdaW(v: V): Promise[W] => ...
+
+async fun async_fun(): Promise[W] =>
+    let promise: Promise[T] = /* obtained from an async call */
+    let t: T = await promise
+    let v: V = await lambdaV(t)
+    let w: W = await lambdaW(v)
+    // from the point of view of the caller,
+    this is a Promise[W] because this "fun" is marked as "async"
+    w 
+```
+
+To:
+
+```pony
+async fun lambdaV(t: T): Promise[V] => ...
+async fun lambdaW(v: V): Promise[W] => ...
+
+fun async_fun(): Promise[W] =>
+    let promise: Promise[T] = /* obtained from an async call */
+    let result: Promise[W] =
+        promise.next[V]({(t: T) => lambdaV(t)})
+            .next[W]({(v: V) => lambdaW(v)})
+    result
+```
```

Comment:
Please show full desugaring. I can't evaluate what desugaring is supposed to be based on this partial desugared example.

Reply:
Can't the compiler do that with the above code once you fill in the blanks?

Reply:
Please show full desugaring.